### ISSUE TITLE
Permalink broken when both 'folder' and 'base' are set

### ISSUE THREAD

**Issue by Kagetsuki** [movabletype/movabletype#150]
There's some sort of conflicting condition that occurs when creating a page through the API. For example, say we have a folder with ID 1 with a path a/b/ and we want to create "index.html" in this folder through the API. The following patterns will produce different results:

folder | basename | resulting folder | resulting permalink | expected result
-------+---------------+---------------+--------------------------+---------------------
a/b    | index          | a/b               |  a/b/post-0.html         |   a/b/index.html
-------+---------------+---------------+--------------------------+---------------------
a/b    |                    | a/b              |  a/b/post-0.html         |   a/b/post-0.html
-------+---------------+---------------+--------------------------+---------------------
         | index          |                    |  index.html               |  index.html
-------+---------------+---------------+--------------------------+---------------------
a/b    |  a/b/index    | a/b             |  a/b/a/b/index.html         |   a/b/index.html
-------+---------------+---------------+--------------------------+---------------------
         |  a/b/index    |                   |  a/b/index.html         |   a/b/index.html

Ideally the first pattern should work, but it does not. The only pattern that works is if you do not set the folder and set the base name as the full path without the .html extension, *but* then you can't sort the file into a folder in the admin screen.

**Comment by yuji** [comment]
@Kagetsuki 

Hello,

Do you mean API as Data API? If yes, please tell me Movable Type Version.

**Comment by Kagetsuki** [comment]
Yup, the Data API. I'm testing on 6.3.6. I have a feeling the internal code works something like this:

1. if folder is set, manage file under folder (automatically generate name from title, if no title or title contains no ASCII characters set title to "post-##") ⇒ set file path and name based on folder and title (basename is not used? and automatically generated)
2. if basename is set, check to make sure the base name does not conflict with folder path (path portion of base name matches folder path) ⇒ if base name conflicts with folder path ignore base name / set it to null
3. permalink is generated as folder path + base name

The logic here probably made sense, because if you had folder path a/b and base name c/d/e.html, folder sorting wouldn't work correctly. The issue is there's no other attribute but base name to set the file name, and in the case of using the API you'd need to set a base name as "index" to get the apex of a folder or you'd need to set manually to some name to fit with links in other pages you're posting.

Ideally if MT would extract folders automatically from the base name, and create folders if they didn't already exist, it would not only solve this issue but would also eliminate the need to pass folder arguments (which require folder ID, meaning you need to get folders from API first and then extrapolate the ID before you can even post a page!).

Thanks for looking into this - please tell me if you need any more info.

**Comment by Kagetsuki** [comment]
I hate to bump but we're looking to finish a site in the next 10 days and without this working it's going to be a bit of a mess. If there's some alternate solution I'd love to hear about it - otherwise do you think we can expect a patch at some point?

**Comment by Kagetsuki** [comment]
> Once a page has been created with a basename of 'index' for example, the above code does not allow for any other pages to have the same 'index' basename, even if the page is in a different folder. In this case, the Data API generates a unique basename ("post-##") instead of allowing "index"

I'd have to double check but that sounds dead on! Specifically line 4 of the table I made in the first post would seem to confirm this:

|folder | basename | resulting folder | resulting permalink | expected result|
|-------|---------------|---------------------|--------------------------|---------------------|
|a/b    |  a/b/index    | a/b             |  a/b/a/b/index.html         |   a/b/index.html|

But, as I noted, permalink generation is folder + basename, so the "resulting permalink" above has the path twice, and sending the permalink as a parameter to the API just gets ignored.

Definitely hoping for a quick fix on the API. The site we're doing is a monster and importing all the pages through the admin panel by hand would probably take months... with the API we have it down to about 10 minutes!

**Comment by Kagetsuki** [comment]
何度も投稿して大変申し訳ありませんが、ライセンス費を払っているお客様に綺麗に整理されているサイトの納品がしたいのです。 @dphillips 様が述べているようにそこまで複雑な問題ではなさそうなので、早めに対応して貰えれば大変ありがたいです。
よろしくお願い致します。

**Comment by yuji** [comment]
@Kagetsuki 
大変ご迷惑をお掛けしております。できる限り、早急の対応を検討します。
